src/marin/post_training/train_worker.py

```diff
@@ -22,477 +22,297 @@
+
+    # KL regularization against reference policy (prevents drift)
+    kl_penalty = jnp.subtract(current_logprobs, reference_logprobs_array)
+    kl_loss = -jnp.sum(kl_penalty * loss_masks_array) / jnp.sum(loss_masks_array)
+
+    loss = reinforce_loss + kl_coef * kl_loss
+
+    with levanter.tracker.defer_tracker_for_jit() as metrics:
+        metrics.update(
+            {
+                "rloo/token_loss": jnp.mean(token_loss),
+                "rloo/log_ratio": jnp.mean(log_ratio),
+                "rloo/reinforce_loss": reinforce_loss,
+                "rloo/loss": loss,
+                "rloo/kl_loss": kl_loss,
+                "rloo/importance_ratio_mean": jnp.mean(ratio),
+                "rloo/importance_ratio_max": jnp.max(ratio),
+            }
+        )
+
+    return loss
+
+
+class StreamingRolloutLoader:
+    """Direct loader for streaming rollout data that bypasses Levanter's DataLoader."""
+
+    def __init__(self, data_loader: ReplayDataLoader, config):
+        """Initialize the streaming rollout loader.
+
+        Args:
+            data_loader: The replay data loader to get batches from
+            config: Trainer config with mesh and axis mapping information
+        """
+        self.data_loader = data_loader
+        self.config = config
+        self.timeout = 60.0
+
+    def __iter__(self):
+        """Yield batches continuously from the replay buffer."""
+        while True:
+            # Get batch from replay buffer
+            batch = self.data_loader.get_training_batch(timeout=self.timeout)
+            if batch is None:
+                # In a real scenario we might want to handle this differently
+                # For now, we'll raise StopIteration if no data is available
+                logger.warning("No batch available from replay buffer after timeout")
+                raise StopIteration("No data available from replay buffer")
+
+            # Convert to JAX with named axes
+            named_batch = self._convert_to_named_batch(batch)
+
+            # Apply sharding if we have a mesh context
+            if hasattr(self.config, "device_mesh"):
+                with self.config.device_mesh:
+                    if hasattr(self.config, "compute_axis_mapping"):
+                        named_batch = hax.shard(named_batch, self.config.compute_axis_mapping)
+
+            yield named_batch
+
+    def _convert_to_named_batch(self, batch: RolloutBatch):
+        """Convert numpy arrays to JAX arrays with proper named axes."""
+        from marin.post_training.rollout_storage import JaxRolloutBatch
+
+        # Convert to JAX arrays first
+        jax_batch = batch.to_jax()
+
+        # Create batch and sequence axes
+        batch_size, seq_len = jax_batch.input_ids.shape
+        Batch = hax.Axis("batch", batch_size)
+        Pos = hax.Axis("position", seq_len)
+
+        # Add named axes to all fields
+        return JaxRolloutBatch(
+            input_ids=hax.named(jax_batch.input_ids, (Batch, Pos)),
```

Comment:
you can just use string names now if you want: hax.named(jax_batch.input_ids, ("batch", "position")),